server/src/main/java/org/elasticsearch/action/ActionListener.java

```diff
@@ -315,12 +315,28 @@ protected void innerOnFailure(Exception e) {
     /**
      * Completes the given listener with the result from the provided supplier accordingly.
      * This method is mainly used to complete a listener with a block of synchronous code.
+     *
+     * If the supplier fails, the listener's onFailure handler will be called.
+     * It is the responsibility of {@code delegate} to handle its own exceptions inside `onResponse` and `onFailure`.
      */
     static <Response> void completeWith(ActionListener<Response> listener, CheckedSupplier<Response, ? extends Exception> supplier) {
+        Response response;
         try {
-            listener.onResponse(supplier.get());
+            response = supplier.get();
         } catch (Exception e) {
-            listener.onFailure(e);
+            try {
+                listener.onFailure(e);
+            } catch (RuntimeException ex) {
+                assert false : ex;
+                throw ex;
+            }
+            return;
+        }
+        try {
+            listener.onResponse(response);
+        } catch (RuntimeException ex) {
```

Comment:
I'm starting to wonder ... maybe (as the eventual goal of our efforts here, don't think it's possible right now), we should just make `ActionListener` an `abstract` class and enforce that nothing is thrown in `onResponse` and `onFailure` by simply making these lines call some `protected` `innerOnResponse` or so.
It seems just enforcing the not throwing in `onComplete` and `map` is simply a small part of the overall issue of enforcing the callbacks to handle their own exceptions and we could just dry things up that way?

Reply:
I think I agree to this, but unfortunately it does not free us from avoiding for instance `wrap`, since it delegates failures from inner calls to listener.onResponse to onFailure in production (i.e., without assertions).